### PR TITLE
Fixes for casper-client-rs #57

### PR DESCRIPTION
**Pull request by bpr**
Merge me!

### CODE REVIEW

File: lib/rpcs/v1_4_5/speculative_exec.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+/// The `result` field of a successful JSON-RPC response to an `speculative_exec` request.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct SpeculativeExecResult {
+    /// The JSON-RPC server version.
+    pub api_version: ProtocolVersion,
+    /// The deploy hash.
+    pub deploy_hash: DeployHash,
+}
```
Comment (Fraser999) [inline lib/rpcs/v1_4_5/speculative_exec.rs:35]: This needs to be the same as the type defined by the node [here](https://github.com/casper-network/casper-node/blob/cc95de3a80cdada1d67a84a2178f23bcf10e36ae/node/src/components/rpc_server/rpcs/speculative_exec.rs#L58-L66):
```suggestion
pub struct SpeculativeExecResult {
    /// The JSON-RPC server version.
    pub api_version: ProtocolVersion,
    /// Hash of the block on top of which the deploy was executed.
    pub block_hash: BlockHash,
    /// Result of the execution.
    pub execution_result: ExecutionResult,
}
```

File: lib/cli.rs [hunk 2]
```diff
@@ -73,6 +73,7 @@ pub use simple_args::help as simple_args_help;
 pub async fn put_deploy(
     maybe_rpc_id: &str,
     node_address: &str,
+    maybe_block_identifier: &str,
```
Comment (Fraser999) [inline lib/cli.rs:76]: As with ordering the args in the subcommands, I'd probably try and keep the rpc_id, node_address and verbosity parameters grouping unbroken.

File: src/deploy/creation_common.rs [hunk 3]
```diff
@@ -235,6 +236,33 @@ pub(super) fn payment_str_params(matches: &ArgMatches) -> PaymentStrParams<'_> {
...
+            .help(ARG_HELP)
+            .display_order(DisplayOrder::BlockIdentifier as usize)
+    }
+
+    pub(in crate::deploy) fn get(matches: &ArgMatches) -> &str {
+        matches
+            .get_one::<String>(ARG_NAME)
+            .map(String::as_str)
+            .unwrap_or_default()
+    }
```
Comment (Fraser999) [inline src/deploy/creation_common.rs:263]: I think we need to change the return type here to represent the three cases: `--speculative-exec` wasn't provided, `--speculative-exec` was provided with no value, and `--speculative-exec` was provided with a value:
```suggestion
    pub(in crate::deploy) fn get(matches: &ArgMatches) -> (bool, &str) {
        let value = matches
            .get_one::<String>(ARG_NAME)
            .map(String::as_str)
            .unwrap_or_default();
        match value {
            "" => (false, ""),
            DEFAULT_MISSING_VALUE => (true, ""),
            _ => (true, value),
        }
    }
```

File: lib/lib.rs [hunk 4]
```diff
@@ -100,6 +101,26 @@ use types::{Deploy, DeployHash, MAX_SERIALIZED_SIZE_OF_DEPLOY};
...
+/// Sends a JSON-RPC `speculative_exec` request to the specified node.
+///
+/// For details of the parameters, see [the module docs](crate#common-parameters).
+pub async fn speculative_exec(
+    rpc_id: JsonRpcId,
+    node_address: &str,
+    maybe_block_identifier: Option<BlockIdentifier>,
+    verbosity: Verbosity,
+    deploy: Deploy,
+) -> Result<SuccessResponse<PutDeployResult>, Error> {
```
Comment (Fraser999) [inline lib/lib.rs:115]: This needs to return a `SpeculativeExecResult` rather than a `PutDeployResult`.

As with other result types, it should be exported via `lib/rpcs/results.rs`.

File: lib/cli.rs
Comment (Fraser999) [inline lib/cli.rs:1]: Since the speculative execution response is different to the put deploy response, we'll need to create three new functions here for this - something like `speculative_put_deploy`, `speculative_send_deploy_file` and `speculative_transfer` I guess.

File: src/deploy/put.rs [hunk 5]
```diff
@@ -33,6 +34,7 @@ impl ClientCommand for PutDeploy {
         let verbosity_level = common::verbose::get(matches);
 
         let secret_key = common::secret_key::get(matches);
+        let maybe_block_identifier = creation_common::speculative_exec::get(matches);
```
Comment (Fraser999) [inline src/deploy/put.rs:37]: With the other changes suggested, this will become `let (speculative_exec, maybe_block_identifier) =` and below we'll call `casper_client::cli::speculative_put_deploy` or `casper_client::cli::put_deploy` depending upon the bool.

**Review by Fraser999 — CHANGES_REQUESTED** [review]

File: lib/lib.rs [hunk 6]
```diff
@@ -100,6 +101,26 @@ use types::{Deploy, DeployHash, MAX_SERIALIZED_SIZE_OF_DEPLOY};
 pub use validation::ValidateResponseError;
 pub use verbosity::Verbosity;
 
+/// Puts a [`SpeculativeExec`] to the network for execution.
```
[the remark this answers is not in the document]
Comment (bpr) [reply lib/lib.rs:104]: Done, yes much better.

File: lib/rpcs/v1_4_5/speculative_exec.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+impl SpeculativeExecParams {
+    pub(crate) fn new(block_identifier: Option<BlockIdentifier>, deploy: Deploy) -> Self {
+        SpeculativeExecParams {
+            block_identifier,
+            deploy,
+        }
+    }
+}
+
+/// The `result` field of a successful JSON-RPC response to an `speculative_exec` request.
```
[the remark this answers is not in the document]
Comment (bpr) [reply lib/rpcs/v1_4_5/speculative_exec.rs:27]: Ha ha, the perils of copy pasta! Thanks!

File: lib/rpcs/v1_4_5/speculative_exec.rs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+use serde::{Deserialize, Serialize};
```
[the remark this answers is not in the document]
Comment (bpr) [reply lib/rpcs/v1_4_5/speculative_exec.rs:1]: Fixed.

File: src/deploy/creation_common.rs [hunk 9]
```diff
@@ -235,6 +236,33 @@ pub(super) fn payment_str_params(matches: &ArgMatches) -> PaymentStrParams<'_> {
...
 
+/// Handles providing the arg for speculative execution.
+pub(super) mod speculative_exec {
+    use super::*;
+
+    const ARG_NAME: &str = "speculative_exec";
+    const ARG_VALUE_NAME: &str = "HEX STRING OR INTEGER";
+    const ARG_HELP: &str =
+        "The endpoint to perform the speculative execution, identified by a BlockIdentifier, \
+        which is either a cryptographic hash or a u64.";
```
[the remark this answers is not in the document]
Comment (bpr) [reply src/deploy/creation_common.rs:247]: That's a pretty long help string. Almost two tweets!

File: src/deploy/creation_common.rs [hunk 10]
```diff
@@ -235,6 +236,33 @@ pub(super) fn payment_str_params(matches: &ArgMatches) -> PaymentStrParams<'_> {
...
+
+    pub(in crate::deploy) fn arg() -> Arg {
+        Arg::new(ARG_NAME)
+            .long(ARG_NAME)
+            .required(false)
+            .value_name(ARG_VALUE_NAME)
```
[the remark this answers is not in the document]
Comment (bpr) [reply src/deploy/creation_common.rs:253]: `    const DEFAULT_MISSING_VALUE: &str = "0";
`

Will that cause problems?

File: lib/rpcs/v1_4_5/speculative_exec.rs (hunk 1 above)
Comment (bpr) [reply lib/rpcs/v1_4_5/speculative_exec.rs:35]: That would add a dependency on casper-node on account of BlockHash. What's the suggested workaround? IIRC, you don't want to introduce dependencies on node.

File: lib/cli.rs (hunk 2 above)
Comment (bpr) [reply lib/cli.rs:76]: I gotta say, if the grouping of parameters is that important, making them into a struct would make sense.

File: lib/lib.rs (hunk 4 above)
Comment (bpr) [reply lib/lib.rs:115]: Done.

File: src/deploy/put.rs (hunk 5 above)
Comment (bpr) [reply src/deploy/put.rs:37]: I took a slightly different but I hope, equivalent path, using an option type to encode the speculative_exec/no_speculative_exec distinction.

File: lib/cli.rs [hunk 11]
```diff
@@ -86,6 +86,27 @@ pub async fn put_deploy(
         .map_err(CliError::from)
 }
 
+/// Speculatively creates a [`Deploy`] and sends it to the network for execution.
```
Comment (Fraser999) [inline lib/cli.rs:89]: I think the term "speculative" applies exclusively to the execution of the deploy on the node receiving it.  It follows quite a different code path through the node, and a bunch of caveats are implied in any successful response.

So while I don't think this is the place to try and describe the node's process, I guess I'd rephrase this to something like:
```suggestion
/// Creates a [`Deploy`] and sends it to the specified node for speculative execution.
```

File: lib/cli.rs [hunk 12]
```diff
@@ -142,6 +163,25 @@ pub async fn send_deploy_file(
         .map_err(CliError::from)
 }
 
+/// speculatively eads a previously-saved [`Deploy`] from a file and sends it to the network for execution.
```
Comment (Fraser999) [inline lib/cli.rs:166]:
```suggestion
/// Reads a previously-saved [`Deploy`] from a file and sends it to the specified node for
/// speculative execution.
```

File: lib/cli.rs [hunk 13]
```diff
@@ -178,6 +218,44 @@ pub async fn transfer(
         .map_err(CliError::from)
 }
 
+/// Speculatively transfers funds between purses.
```
Comment (Fraser999) [inline lib/cli.rs:221]:
```suggestion
/// Creates a [`Deploy`] to transfer funds between purses, and sends it to the specified node for
/// speculative execution.
```

**Review by Fraser999 — CHANGES_REQUESTED** [review]
Nearly there!  I made a few suggestions, but there's just one blocker left (the issue in `speculative_exec::get()` in creation_common.rs).

File: src/deploy/creation_common.rs (hunk 3 above)
Comment (bpr) [reply src/deploy/creation_common.rs]: I don't understand why it doesn't work with

```Rust
    const DEFAULT_MISSING_VALUE: &str = "";
```

If that gets put in why isn't it `Some("")`?

Anyways, I changed as per your suggestion.

File: lib/cli.rs [hunk 14]
```diff
@@ -86,6 +86,27 @@ pub async fn put_deploy(
         .map_err(CliError::from)
 }
 
+/// Speculatively creates a [`Deploy`] and sends it to the network for execution.
+///
+/// For details of the parameters, see [the module docs](crate::cli#common-parameters) or the docs
+/// of the individual parameter types.
+pub async fn speculative_put_deploy(
+    maybe_speculative_exec: &str,
```
[the remark this answers is not in the document]
Comment (bpr) [reply lib/cli.rs:94]: Done

File: lib/cli.rs (hunk 13 above)
Comment (bpr) [reply lib/cli.rs:221]: Done

File: lib/cli.rs (hunk 12 above)
Comment (bpr) [reply lib/cli.rs:166]: Done

File: lib/cli.rs (hunk 11 above)
Comment (bpr) [reply lib/cli.rs:89]: Done.

**Review by Fraser999 — APPROVED** [review]
LGTM now, but I'd be happy for you to change the `speculative_exec::get()` function one last time as per the comment.  Up to you - it's functionally correct now - just more verbose than need be.

**Comment by bpr** [comment]
What is an appropriate test for this PR?

**Review by darthsiroftardis — APPROVED** [review]
